cfgrammar/src/lib/span.rs

```diff
@@ -60,25 +60,29 @@ impl Default for NewlineToLineColCache {
 
 impl NewlineToLineColCache {
     /// Feed more input into the cache, calculating newlines data from it.
-    /// The `src` string given is treated as though it is contiguous with previous a call to `feed`.
-    /// subsequent logical lines, and logical bytes reflecting the sum of total of offsets which
-    /// have been fed.
+    /// The `src` string given is treated as though they were the concatenation of strings given
+    /// by previous calls to `feed`. Subsequent logical lines, and bytes given relative
+    /// to that cumulative string.
     pub fn feed(&mut self, src: &str) {
-        let start_pos = self.newlines[self.newlines.len() - 1] + self.trailing_bytes;
+        let start_pos = self.newlines.last().unwrap() + self.trailing_bytes;
         self.newlines
             .extend(src.char_indices().filter_map(|c| match c {
-                (offset, '\n') | (offset, '\r') => {
+                (offset, '\n') => {
                     self.trailing_bytes = 0;
                     Some(start_pos + offset + 1)
                 }
+                (_, '\r') => {
```

Comment:
Without multiple tests to tell what this actually does, I'm not super happy with this: it looks like it's going to have some weird consequences. But I might be wrong! So let's add tests please :)

Reply:
Yeah, you are correct here this is going to mess up byte offsets, and if we aren't dealing with `\r` by itself the corner case I was imagining won't actually be a case were going to handle.

The thing I was worried about is '\r' contributing to the column calculation, in particular if we have input like:
`cache.feed("aaaa\nbbbb"); cache.feed("mnop"); cache.byte_to_line_and_column(.., "mnop", ...);`
we need to (but aren't currently) dealing with the need for the characters "bbbb" to contribute to the column count.

to deal with this case we need *both* a `trailing_bytes` and `trailing_chars`, and `\r` *should* contribute to `trailing_bytes` so as to avoid messing up byte offsets.  But it seems like it should *not* contribute to `trailing_chars`. If we were dealing with `\r` by itself and `\r\n`, this would seem to be much more problematic.  As then we get cases like `cache.feed("\r"); cache.feed("\n");`.

At least I feel like I understand the case that lead me to make this change now, so i'll get to work on codifying them as tests...

Reply:
So it's actually worse than needing `trailing_chars` I believe the tests I've added need a `Vec<usize>` i.e. a vector of trailing chars one for each call to `feed`, or perhaps a `HashMap<usize, usize>` to map `src_pos` to `trailing_chars`.

This is seeming to get a bit hairy so i am going to wait to proceed in that until you have a chance to weigh in.

Reply:
Well I just went ahead and implemented the above in 0f8f14f so we could at least look at the damage before deciding...